astropy/io/misc/pyarrow/csv.py

```diff
@@ -0,0 +1,402 @@
+        Whether newline characters are allowed in CSV values. Setting this to True
+        reduces the performance of multi-threaded CSV reading.
+
+    Returns
+    -------
+    Table
+        An astropy Table containing the data from the CSV file.
+    """
+    _, csv = get_pyarrow_csv()
+
+    parse_options = csv.ParseOptions(
+        delimiter=delimiter,
+        quote_char=quotechar,
+        double_quote=doublequote,
+        escape_char=escapechar,
+        newlines_in_values=newlines_in_values,
+    )
+
+    read_options = get_read_options(header_start, data_start, names, encoding)
+    convert_options = get_convert_options(include_names, dtypes, null_values)
+
+    if comment is not None:
+        input_file_stripped, header_start_new = strip_comment_lines(
+            input_file, comment, encoding, header_start, data_start
+        )
+        if header_start_new is not None:
+            header_start = header_start_new
+        if input_file_stripped is not None:
+            input_file = input_file_stripped
+
+    table_pa = csv.read_csv(
+        input_file,
+        parse_options=parse_options,
+        read_options=read_options,
+        convert_options=convert_options,
+    )
+    table_apt = convert_pa_table_to_astropy_table(table_pa)
+
+    return table_apt
+
+
+def get_pyarrow_csv():
+    """Helper function to import pyarrow and pyarrow.csv."""
+    if not HAS_PYARROW:
+        raise ModuleNotFoundError(
+            "pyarrow is required to read and write pyarrow.csv files"
+        )
+    import pyarrow as pa
+    from pyarrow import csv
+
+    return pa, csv
+
+
+def convert_pa_array_to_numpy(
+    arr: Union["pa.Array", "pa.ChunkedArray"],
+) -> "npt.NDArray":
+    """
+    Convert a PyArrow array to a NumPy array.
+
+    Parameters
+    ----------
+    arr : pyarrow.Array, pyarrow.ChunkedArray
+        The PyArrow array to be converted.
+
+    Returns
+    -------
+    np.ndarray or np.ma.MaskedArray
+        The converted NumPy array. If the input array contains null values, a masked
+        array is returned with nulls masked out.
+
+    Notes
+    -----
+    - If the input array is of string type, it delegates the conversion to
+      `convert_pa_string_array_to_numpy`.
+    - If the input array contains null values, they are replaced with 0 before
+      conversion, and a masked array is returned with the null positions masked.
+    - If the input array does not contain null values, the result is an ndarray view of
+      the underlying data.
+    """
+    import pyarrow.compute as pc
+
+    is_string = arr.type == "string"
+
+    if arr.null_count == 0:
+        # No nulls, just return an ndarray view of the pyarray
+        if is_string:
+            max_length = pc.max(pc.utf8_length(arr))
+            out = np.empty(len(arr), dtype=f"U{max_length}")
+            out[:] = arr.to_numpy()
+        else:
+            out = arr.to_numpy()
+    else:
+        # Return a Masked Array copy, with nulls filled with zero.
```

Comment:
Do we care what is underneath masked values? For numerical data at least, I think not, which means we can just have a masked array view.
Alternatively, can we just *set* the values to zero? It is not as if the pa array somehow needs to be preserved. Or are pa arrays read-only?

Reply:
I'm definitely not an expert on pyarrow, but it does appear that the array is read-only:
```
In [1]: import pyarrow as pa
   ...: arr = pa.array([1, 2, None, 3], type=pa.int8())
   ...: fill_value = pa.scalar(5, type=pa.int8())

In [2]: arr
Out[2]: 
<pyarrow.lib.Int8Array object at 0x10e21b8e0>
[
  1,
  2,
  null,
  3
]

In [3]: arr[arr.is_null()] = fill_value
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[3], line 1
----> 1 arr[arr.is_null()] = fill_value

TypeError: 'pyarrow.lib.Int8Array' object does not support item assignment
```

About the "masked array view" comment, I'm not sure what you meant by that.